### ISSUE TITLE
Add a Replace() function

### ISSUE THREAD

**Issue by aloisdg** [morelinq/MoreLINQ#145]
Hello,

For a small project, I needed a replace function. I wrote one. If it is judged useful, I will write a proper PR. You can try the code in [.NETFiddle](https://dotnetfiddle.net/1uZXeQ).

```C#
using System;
using System.Collections.Generic;
using System.Linq;

public static partial class MoreEnumerable
{
	// TSource is not found
	
	public static IEnumerable<T> Replace<T>(this IEnumerable<T> source, T old, T neo)
	{
       	if (source == null)
                throw new ArgumentNullException("source");
			 
		foreach (var item in source)
		{
			if (item.Equals(old))
				yield return neo;
			else
				yield return item;
		}
	}
	
	public static IEnumerable<T> ReplaceWhere<T>(this IEnumerable<T> source, Func<T, bool> condition, T neo)
	{
       	if (source == null)
                throw new ArgumentNullException("source");
			 
		foreach (var item in source)
		{
			if (condition(item))
				yield return neo;
			else
				yield return item;
		}
	}
	
	public static IEnumerable<T> ReplaceFirst<T>(this IEnumerable<T> source, T old, T neo)
	{
		if (source == null)
        	throw new ArgumentNullException("source");
		
		var isFirst = true;
		foreach (var item in source)
		{
			if (item.Equals(old) && isFirst)
			{
				isFirst = false;
				yield return neo;
			}
			else
				yield return item;
		}
	}

	// I am not proud of this one but it is just to show the idea.
	public static IEnumerable<T> ReplaceLast<T>(this IEnumerable<T> source, T old, T neo)
	{
		return source.Reverse().ReplaceFirst(old, neo).Reverse();
	}
}

public class Program
{
	public static void Main()
	{
		var enumerable = Enumerable.Range(0, 10);

		enumerable = enumerable.Replace(6, 64);
		enumerable = enumerable.ReplaceWhere(x => x % 2 != 0, 9000);
		enumerable = enumerable.ReplaceFirst(9000, 1);
		enumerable = enumerable.ReplaceLast(9000, 9);
		
		foreach (var item in enumerable)
			Console.WriteLine(item);
	}
}
```

Cheers

**Comment by atifaziz** [comment]
Thanks for your proposal but I'm not sure how `Replace` and `ReplaceWhere` are any different than `Select`:

```c#
// enumerable = enumerable.Replace(6, 64)
enumerable = enumerable.Select(x => x == 6 ? 64 : x)
// enumerable = enumerable.ReplaceWhere(x => x % 2 != 0, 9000)
enumerable = enumerable.Select(x => x % 2 != 0 ? 9000 : x);
```

In fact, that's how they can be implemented internally.

I can see `Replace` just being a way of simplifying the `Select` call when you're just looking to replace one value with another but it would need several overloads before it can be made more generally useful. For example, for strings especially, you would want to be able to supply an equality comparer. That's a good idea anyhow since it would help with handling nulls, something that your `Replace` doesn't appear to handle but would work if you just express it as a `Select` projection (because the `==` takes care of that but `item.Equals(old)` doesn't).

`ReplaceFirst` could be seen as a `Replace` overload taking a count of replacements and supplying a value of 1 will mean the same as *replace first and only*.

`ReplaceLast` has some merit given that it can be complex to implement but I would say that for it to have any utility, it should at least be optimized for lists so that it doesn't need reverse twice if the source can be indexed into directly.

To illustrate some of my points, I quickly prototyped a `Replace` using [`Scan`][scan] and [`Pairwise`][pairwise] that are already part of MoreLINQ: 

```c#
public static IEnumerable<T> Replace<T>(this IEnumerable<T> source,
    T search, T replacement, int count,
    IEqualityComparer<T> comparer = null)
{
    comparer = comparer ?? EqualityComparer<T>.Default;

    if (count == 0)
        return source;
    
    var result =
        (count < 0 ? source.Reverse() : source)
        .Scan(new { Count = 0, Item = default(T) },
              (s, e) => new { Count = s.Count + (comparer.Equals(e, search) ? 1 : 0),
                              Item  = e })
        .Pairwise((prev, curr) => prev.Count < curr.Count 
                                  && curr.Count <= Math.Abs(count)
                                ? replacement 
                                : curr.Item);
 
    return count < 0 ? result.Reverse() : result;
}
```

The above (I think) handles all cases in one implementation. For the same behavior as `ReplaceLast`, you simply supply a negative count. To replace all occurrences, the `count` should be `int.MaxValue` though I would make that optional through an overload and not assume any maximum.

In summary, I would say the following needs addressing:

- [ ] Equality comparer argument
- [ ] Maximum number of replacements (count) argument
- [ ] Possibly also a starting occurrence argument, which enables *n-th* or subset replacement
- [ ] Null handling
- [ ] Deferred execution
- [ ] Optimization for lists
- [ ] Optionality via overloads
- [ ] Unit tests with maximum coverage

If you're happy to take all of that in a PR then it would be a welcome addition. If this is more than you were planning to invest as energy then feel free to close the issue and revisit in the future.

  [scan]: https://morelinq.github.io/1.x/ref/api/html/M_MoreLinq_MoreEnumerable_Scan__2.htm
  [pairwise]: https://morelinq.github.io/1.x/ref/api/html/M_MoreLinq_MoreEnumerable_Pairwise__2.htm

**Comment by aloisdg** [comment]
Great answer. Thank you!

Yes sure I can look to do all of that. I cannot take the time to do it right now. 
I think I will close it for now, then I will open it later.
